interfaces: refactor: move `Mining` and `BlockTemplate`  implementation to miner

ismaelsadeeq opened the pull request:
#### Motivation

In  [Internal interface guidelines](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#internal-interface-guidelines)

It's stated that 

> Interface method definitions should wrap existing functionality instead of implementing new functionality. Any substantial new node or wallet functionality should be implemented in [src/node/](https://github.com/bitcoin/bitcoin/blob/master/src/node) or [src/wallet/](https://github.com/bitcoin/bitcoin/blob/master/src/wallet) and just exposed in [src/interfaces/](https://github.com/bitcoin/bitcoin/blob/master/src/interfaces) instead of being implemented there, so it can be more modular and accessible to unit tests.

However the some methods in the newly added  `BlockTemplateImpl` and `MinerImpl`  classes partially enforces this guideline, as the implementations of the `submitSolution`, `waitNext`, and `waitTipChanged` methods reside within the class itself.

#### What the PR Does

This PR introduces a simple refactor by moving certain method implementations from `BlockTemplateImpl` into the miner module. It introduces three new functions:

1.  Remove rundundant coinbase fee check in `waitNext`
2. **`AddMerkleRootAndCoinbase`**: Computes the block's Merkle root, inserts the coinbase transaction, and sets the Merkle root in the block. This function is called by `submitSolution` before the block is submitted for processing.

3. **`WaitAndCreateNewBlock`**: Returns a new block template either when transaction fees reach a certain threshold or when a new tip is detected. If a timeout is reached, it returns `nullptr`. The `waitNext` method in `BlockTemplateImpl` now simply wraps this function.
4. Move `GetTip` implementation to miner.

5. **`WaitTipChanged`**: Returns the tip when the chain it changes, or `nullopt` if a timeout or interrupt occurs. The `waitTipChanged` method in `MinerImpl` now calls `GetTip` after invoking `ChainTipChanged`, and returns the tip.

#### Behavior Change

- We now only `Assert` for  a valid chainman and notifications pointer once.

ismaelsadeeq committed:
interfaces: refactor: move `submitSolution` implementation to miner

- Create a new function `AddMerkleRootAndCoinbase` that compute the
  block's merkle root, insert the coinbase transaction and the merkle
  root into the block.

ismaelsadeeq committed:
interfaces: refactor: move `waitNext` implementation to miner

- We now assert for a valid chainman and notifications once when
 invoking WaitAndCreateNewBlock function.

ismaelsadeeq committed:
interfaces: refactor: move `waitTipChanged` implementation to miner

- This commit creates a function `WaitTipChanged` that waits for the connected
  tip to change until timeout elapsed.

- This function is now used by `waitTipChanged`

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -446,4 +449,94 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
     block.hashMerkleRoot = BlockMerkleRoot(block);
 }
+
+std::optional<std::unique_ptr<CBlockTemplate>> WaitNext(NodeContext& node, const uint256& prev_block_hash, const std::vector<CAmount>& tx_fees, const BlockWaitOptions& options, const BlockAssembler::Options& assemble_options)
```
In commit "interfaces: refactor: move `waitNext` implementation to miner" (a5145ba0fc3339392ff71bf20613812e0d8c8efc)

This change generally looks good, but I would suggest tweaking this commit and the next one to pass Chainman& and KernelNotification& parameters directly instead of NodeContext& because NodeContext exposes a lot of node state that this code shouldn't need access to and also brings in dependencies this code shouldn't need to depend on.

ismaelsadeeq replied:
Thanks for taking a look
I've taken your suggestion with the addition of `CTxMemPool*` as a parameter to `WaitNext`.
I think it's best to pass it as well, to make the method self contained and allow for testing with a custom mempool.

Also, if I use the mempool pointer in the active chainstate not the node, a miner test fails.
This happens because in `TestPackageSelection`, we construct a new mempool and update the node interfaces' mempool pointer, but we don't update the raw pointer stored in the active chainstate.

ismaelsadeeq commented:
- Updated to fix recent comment by @ryanofsky  see [5fa54e2..c8a3fab](https://github.com/bitcoin/bitcoin/compare/77949ef5267e4de4e68a19560bab81e8d5fa54e2..c8a3fabe4090fa2b9a0e8cef73ed5365e8221ad6)

- Also fixed tidy issues

ryanofsky commented on src/node/miner.h:
```diff
@@ -232,6 +234,10 @@ void ApplyArgsManOptions(const ArgsManager& gArgs, BlockAssembler::Options& opti
 void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t version, uint32_t timestamp, uint32_t nonce);
+
+/* Return a new block template when fees rise to a certain threshold or after a new tip; return nullopt if timeout is reached. */
+std::optional<std::unique_ptr<CBlockTemplate>> WaitNext(ChainstateManager& chainman, KernelNotifications& kernel_notifications, CTxMemPool* mempool, const uint256& prev_block_hash,
```
In commit "interfaces: refactor: move `waitNext` implementation to miner" (d11db291d492785125dc0971ed8eaf0a4c981932)

Would be better to return `std::unique_ptr<CBlockTemplate>` instead of `std::optional<std::unique_ptr<CBlockTemplate>>` here because unique_ptr is already nullable.

Also `node::WaitNext` is kind of a vague name. Maybe a name like `WaitAndCreateNewBlock` could be more descriptive. Or maybe just `CreateNextBlock`. Could be other ideas.

ismaelsadeeq replied:
Fixed

ryanofsky commented on src/node/miner.h:
```diff
@@ -238,6 +238,9 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
                                                         const std::vector<CAmount>& tx_fees, const BlockWaitOptions& options, const BlockAssembler::Options& assemble_options);
+
+/* Waits for the connected tip to change until timeout has elapsed. During node initialization, this will wait until the tip is connected (regardless of `timeout`). */
+bool ChainTipChanged(ChainstateManager& chainman, KernelNotifications& kernel_notifications, const uint256& current_tip, MillisecondsDouble& timeout);
```
In commit "interfaces: refactor: move `waitTipChanged` implementation to miner" (c8a3fabe4090fa2b9a0e8cef73ed5365e8221ad6)

I think name `ChainTipChanged` could be a little misleading since this returns a bool and you might think it returns true when the tip has changed, false otherwise. But in reality it returns true whether or not this tip has changed (after the timeout), and the only case it returns false is when the node is shutting down.

To be less confusing I think it would be good to rename this and return the actual tip instead of a bool. Could do:

```diff
```diff
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -974,7 +974,7 @@ public:
 
     std::optional<BlockRef> waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
     {
-        if (ChainTipChanged(chainman(), notifications(), current_tip, timeout)) return getTip();
+        if (WaitTipChanged(chainman(), notifications(), current_tip, timeout)) return getTip();
         return {};
     }
 
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -539,7 +539,7 @@ std::optional<std::unique_ptr<CBlockTemplate>> WaitNext(ChainstateManager& chain
     return std::nullopt;
 }
 
-bool ChainTipChanged(ChainstateManager& chainman, KernelNotifications& kernel_notifications, const uint256& current_tip, MillisecondsDouble& timeout)
+std::optional<uint256> WaitTipChanged(ChainstateManager& chainman, KernelNotifications& kernel_notifications, const uint256& current_tip, MillisecondsDouble& timeout)
 {
     Assume(timeout >= 0ms); // No internal callers should use a negative timeout
     if (timeout < 0ms) timeout = 0ms;
@@ -554,16 +554,14 @@ bool ChainTipChanged(ChainstateManager& chainman, KernelNotifications& kernel_no
         kernel_notifications.m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
             return kernel_notifications.TipBlock() || chainman.m_interrupt;
         });
-        if (chainman.m_interrupt) return false;
+        if (chainman.m_interrupt) return {};
         // At this point TipBlock is set, so continue to wait until it is
         // different then `current_tip` provided by caller.
         kernel_notifications.m_tip_block_cv.wait_until(lock, deadline, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
             return Assume(kernel_notifications.TipBlock()) != current_tip || chainman.m_interrupt;
         });
+        if (chainman.m_interrupt) return {};
+        return Assume(kernel_notifications.TipBlock());
     }
-
-    if (chainman.m_interrupt) return false;
-
-    return true;
 }
 } // namespace node
--- a/src/node/miner.h
+++ b/src/node/miner.h
@@ -239,8 +239,8 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
 std::optional<std::unique_ptr<CBlockTemplate>> WaitNext(ChainstateManager& chainman, KernelNotifications& kernel_notifications, CTxMemPool* mempool, const uint256& prev_block_hash,
                                                         const std::vector<CAmount>& tx_fees, const BlockWaitOptions& options, const BlockAssembler::Options& assemble_options);
 
-/* Waits for the connected tip to change until timeout has elapsed. During node initialization, this will wait until the tip is connected (regardless of `timeout`). */
-bool ChainTipChanged(ChainstateManager& chainman, KernelNotifications& kernel_notifications, const uint256& current_tip, MillisecondsDouble& timeout);
+/* Waits for the connected tip to change until timeout has elapsed. During node initialization, this will wait until the tip is connected (regardless of `timeout`). Returns the current tip hash, or nullopt if the node is shutting down. */
+std::optional<uint256> WaitTipChanged(ChainstateManager& chainman, KernelNotifications& kernel_notifications, const uint256& current_tip, MillisecondsDouble& timeout);
 } // namespace node
 
 #endif // BITCOIN_NODE_MINER_H
```

ryanofsky reviewed (approved):
Code review ACK c8a3fabe4090fa2b9a0e8cef73ed5365e8221ad6

Thanks for the cleanup! This change should make the node interfaces.cpp more managable and organize code better. I left some suggestions for renames and API tweaks but they are not very important.

ismaelsadeeq commented:
Forced pushed to address recent comments see [c8a3fabe..31e3808d](https://github.com/bitcoin/bitcoin/compare/c8a3fabe4090fa2b9a0e8cef73ed5365e8221ad6..31e3808df9c59d36a07cad07df89ae1bf9e63000)

Sjors commented on src/node/miner.h:
```diff
@@ -229,6 +229,9 @@ void RegenerateCommitments(CBlock& block, ChainstateManager& chainman);
 /** Apply -blockmintxfee and -blockmaxweight options from ArgsManager to BlockAssembler options. */
 void ApplyArgsManOptions(const ArgsManager& gArgs, BlockAssembler::Options& options);
+
+/* Compute the block's merkle root, insert the coinbase transaction and the merkle root into the block */
```
In d3f4e64d416aa29d4448dffb4d47e52997b40271 "interfaces: refactor: move `submitSolution` implementation to miner": maybe say "insert or replace"

It's unlikely for `vtx` to be empty, since we we add a coinbase during template creation.

More importantly, it's possible to call `submitSolution()` multiple times, with each call updating the coinbase. This could happen as a race condition or because the first solution was somehow incorrect.

ismaelsadeeq replied:
Done

Sjors commented on src/node/miner.h:
```diff
@@ -232,6 +234,10 @@ void ApplyArgsManOptions(const ArgsManager& gArgs, BlockAssembler::Options& opti
+
+/* Return a new block template when fees rise to a certain threshold or after a new tip; return nullopt if timeout is reached. */
+std::unique_ptr<CBlockTemplate> WaitAndCreateNewBlock(ChainstateManager& chainman, KernelNotifications& kernel_notifications, CTxMemPool* mempool, const uint256& prev_block_hash,
+                                                      const std::vector<CAmount>& tx_fees, const BlockWaitOptions& options, const BlockAssembler::Options& assemble_options);
```
In 497996c3e572c8992d34e54fd13d92984e9ec21c "interfaces: refactor: move `waitNext` implementation to miner" nit, a little less wide:

```diff
diff --git a/src/node/miner.h b/src/node/miner.h
index e20cab0d08..04b12f065e 100644
--- a/src/node/miner.h
+++ b/src/node/miner.h
@@ -235,9 +235,17 @@ void ApplyArgsManOptions(const ArgsManager& gArgs, BlockAssembler::Options& opti
 /* Compute the block's merkle root, insert the coinbase transaction and the merkle root into the block */
 void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t version, uint32_t timestamp, uint32_t nonce);

-/* Return a new block template when fees rise to a certain threshold or after a new tip; return nullopt if timeout is reached. */
-std::unique_ptr<CBlockTemplate> WaitAndCreateNewBlock(ChainstateManager& chainman, KernelNotifications& kernel_notifications, CTxMemPool* mempool, const uint256& prev_block_hash,
-                                                      const std::vector<CAmount>& tx_fees, const BlockWaitOptions& options, const BlockAssembler::Options& assemble_options);
+/**
+ * Return a new block template when fees rise to a certain threshold or after a
+ * new tip; return nullopt if timeout is reached.
+ */
+std::unique_ptr<CBlockTemplate> WaitAndCreateNewBlock(ChainstateManager& chainman,
+                                                      KernelNotifications& kernel_notifications,
+                                                      CTxMemPool* mempool,
+                                                      const uint256& prev_block_hash,
+                                                      const std::vector<CAmount>& tx_fees,
+                                                      const BlockWaitOptions& options,
+                                                      const BlockAssembler::Options& assemble_options);

 /* Waits for the connected tip to change until timeout has elapsed. During node initialization, this will wait until the tip is connected (regardless of `timeout`).
```

ismaelsadeeq replied:
Done!

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -974,32 +974,8 @@ class MinerImpl : public Mining
-        // Must release m_tip_block_mutex before getTip() locks cs_main, to
-        // avoid deadlocks.
-        return getTip();
+        if (WaitTipChanged(chainman(), notifications(), current_tip, timeout)) return getTip();
```
In 31e3808df9c59d36a07cad07df89ae1bf9e63000 "interfaces: refactor: move `waitTipChanged` implementation to miner": this neatly solves the deadlock issue, but I think we should preserve that comment.

Additionally we should explain why `getTip()` is called here, ignoring the block hash returned by `WaitTipChanged`. Perhaps the latter should just return a boolean?

Sjors replied:
Regarding the latter, I see @ryanofsky's point: https://github.com/bitcoin/bitcoin/pull/32378#discussion_r2080452135

But we should still explain why we ignore the result.

ismaelsadeeq replied:
Added the comment as requested.

Sjors reviewed (commented):
Concept ACK

Mostly happy with 31e3808df9c59d36a07cad07df89ae1bf9e63000

ismaelsadeeq committed:
interfaces: remove redundant coinbase fee check in `waitNext`

- vTxFees now does not include the negative coinbase fee,
  hence this check can be removed.

ismaelsadeeq commented:
Force-pushed to address recent comments by @Sjors:

1. Added an explanation for why we ignore the result of `WaitTipChanged`.
2. Clarified the comment for `AddMerkleRootAndCoinbase`.
3. Shortened line lengths for improved readability.

Additionally, in the recent push, I added a commit that skips the fee check during block template total fees computation in `waitNext`. This check is redundant after #31897.

musaHaruna reviewed (commented):
Code review ACK [5134492](https://github.com/bitcoin/bitcoin/pull/32378/commits/51344920ebe1772d8dd3bece789a6510d774ab59)

Code review: After reading through [internal-interface-guidelines](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#internal-interface-guidelines), the new code change adheres to interface guideline and correctly abracts away implementation details to the appropraite files. Naming of function is very descriptive which is good. 

I compiled each commit successfully and ran all mining related functional test successfully. To make sure the code refactor is not breaking functionalities

Sjors committed:
interfaces: move getTip implementation to miner

ismaelsadeeq commented:
@Sjors I've taken your suggestion in this push [0d774ab59..62fc42d475](https://github.com/bitcoin/bitcoin/compare/51344920ebe1772d8dd3bece789a6510d774ab59..62fc42d475df4f23bd93313f95ee7b7eb0d4683f)

If the `GetTip` method from the mining interface isn't going to be used by the TP, it can be deleted as well?

Sjors reviewed (approved):
ACK 62fc42d475df4f23bd93313f95ee7b7eb0d4683f

We could do another round of interface pruning like #31196 closer to the v30 branch-off.

Although `getTip()` isn't used by the TP (anymore), it does seem generally useful and trivial to maintain. But we can decide later if that's a good enough reason.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -1015,16 +1015,12 @@ class BlockTemplateImpl : public BlockTemplate
                 if (current_fees == -1) {
                     current_fees = 0;
                     for (CAmount fee : m_block_template->vTxFees) {
-                        // Skip coinbase
```
In commit "interfaces: remove redundant coinbase fee check in `waitNext`" (02d4bc776bbe002ee624ec2c09d7c3f981be1b17)

Might be help to mention #31897 in the commit description to clarify what "now does not include" refers to

ryanofsky commented on src/node/miner.h:
```diff
@@ -232,6 +234,17 @@ void ApplyArgsManOptions(const ArgsManager& gArgs, BlockAssembler::Options& opti
+std::unique_ptr<CBlockTemplate> WaitAndCreateNewBlock(ChainstateManager& chainman,
+                                                      KernelNotifications& kernel_notifications,
+                                                      CTxMemPool* mempool,
+                                                      const std::unique_ptr<CBlockTemplate>& block_template,
```
In commit "interfaces: refactor: move `waitNext` implementation to miner" (720f201e652885b9e0aec8e62a1bf9590052b320)

Would be good to change `const std::unique_ptr<CBlockTemplate>& block_template` to just `const CBlockTemplate& block_template` since it doesn't look like it's allowed to be null.

ismaelsadeeq replied:
yeah it is not, I will do if I have to retouch.

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -547,4 +547,33 @@ std::optional<BlockRef> GetTip(ChainstateManager& chainman)
+    }
+    if (chainman.m_interrupt) return {};
+
+    // Must release m_tip_block_mutex before getTip() locks cs_main, to
```
In commit "interfaces: refactor: move `waitTipChanged` implementation to miner" (62fc42d475df4f23bd93313f95ee7b7eb0d4683f)


Not sure, but I think it might be a good idea to drop the 4th commit c39ca9d4f7bc9ca155692ac949be2e61c0598a97 and instead change the 5th commit 62fc42d475df4f23bd93313f95ee7b7eb0d4683f to use the `kernel_notifications.TipBlock()` consistently instead switching to `chainman.ActiveChain().Tip()` at the end:

<details><summary>diff</summary>
<p>

```diff
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -554,6 +554,7 @@ std::optional<BlockRef> WaitTipChanged(ChainstateManager& chainman, KernelNotifi
     if (timeout < 0ms) timeout = 0ms;
     if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
     auto deadline{std::chrono::steady_clock::now() + timeout};
+    uint256 new_tip;
     {
         WAIT_LOCK(kernel_notifications.m_tip_block_mutex, lock);
         // For callers convenience, wait longer than the provided timeout
@@ -569,11 +570,12 @@ std::optional<BlockRef> WaitTipChanged(ChainstateManager& chainman, KernelNotifi
         kernel_notifications.m_tip_block_cv.wait_until(lock, deadline, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
             return Assume(kernel_notifications.TipBlock()) != current_tip || chainman.m_interrupt;
         });
+        if (chainman.m_interrupt) return {};
+        new_tip = *kernel_notifications.TipBlock();
     }
-    if (chainman.m_interrupt) return {};
 
-    // Must release m_tip_block_mutex before getTip() locks cs_main, to
-    // avoid deadlocks.
-    return GetTip(chainman);
+    // Must release m_tip_block_mutex before locking cs_main to avoid deadlocks.
+    LOCK(::cs_main);
+    return BlockRef{new_tip, Assert(chainman.m_blockman.LookupBlockIndex(new_tip))->nHeight};
 }
 } // namespace node
```
</p>
</details>

Could also `s/getTip/GetTip()/` if keeping current code

Sjors replied:
As long as getTip() is in the interface it seems consistent with the rest of this PR to move the implementation, so no need to drop c39ca9d4f7bc9ca155692ac949be2e61c0598a97?

Meanwhile you suggestion can be still be used in the 5th commit.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/32378#discussion_r2089392199

> As long as getTip() is in the interface it seems consistent with the rest of this PR to move the implementation, so no need to drop [c39ca9d](https://github.com/bitcoin/bitcoin/commit/c39ca9d4f7bc9ca155692ac949be2e61c0598a97)?

Yes that's fine since it's possible the function could be used other places. I just suggested dropping the commit to shrink the PR since getTip method is already pretty simple and doesn't necessarily need to to be split up. But any approach, including current approach seems pretty good to me, just wanted to raise this idea.

> Moving the implementation might also allow use to annotate the cs_main lock? (I wasn't sure how to deal with it)

Makes sense if there are other callers of GetTip function they could probably benefit from lock annotations that getTip wouldn't provide (since there probably isn't a good solution for annotating virtual methods).

ryanofsky reviewed (approved):
Code review ACK 62fc42d475df4f23bd93313f95ee7b7eb0d4683f. Lots of suggest suggest changes made since last review, altering function names and signatures and also adding new commit to drop negative fee handling. I like the idea of making the wait function return a BlockRef, that is clearer than what I suggested. Left some comments below but they are not important and this looks good as-is
